scheduler/conf.c

```diff
@@ -1043,32 +1054,6 @@ cupsdReadConfiguration(void)
     }
   }
 
- /*
```

Comment:
OK, we don't want to remove this check because a cupsd without listeners cannot be used and that is a fatal condition.

Reply:
The check is moved into ```service_checkin()```, which should cover both cases - if there are no valid listeners in cupsd.conf and if there are no listeners from outer system (like systemd, launchd, upstart).

I've checked the code once more and what I see as redundant (in case there is no listener) is loading all data from disk - but this happens in case cups.sock is set as only listener as well.

I would not like to have the same code on two places, but moving service_checkin and service_checkout into cupsdReadConfiguration() does not make sense as well...

Reply:
And it would free us the ifdef-fu :)

Reply:
But service_checkin is only used when a) you build with launchd/systemd/upstart, and b) when you run cupsd with the "I'm running from launchd/systemd/upstart" option. So we need this check in both places to make sure we don't end up with a non-functional instance of cupsd.

Reply:
It does not seem to be the case - IIUC it is called only on one place in main.c:

```
 588  /*
 589   * If we were started on demand by launchd or systemd get the listen sockets
 590   * file descriptors... 
 591   */
 592 
 593   service_checkin();    
 594   service_checkout(0);
```

this is not guarded by ifdef or OnDemand option. Then the function itself has most of code guarded by OnDemand and ifdefs, but the last check for Listeners is unguarded:

```
2038 #endif /* HAVE_LAUNCHD */
2039 
2040   if (cupsArrayCount(Listeners) == 0)
2041   {
2042    /*
2043     * No listeners!
2044     */
2045 
2046     cupsdLogMessage(CUPSD_LOG_EMERG, "service_checkin: No listener sockets present.");
2047 
2048    /*
2049     * Commit suicide...
2050     */
2051 
2052     cupsdEndProcess(getpid(), 0);
2053   }
```

thus, since ```service_checkin()``` does not seem guarded in ```main()```, the check is called every time cupsd is loaded, regardless whether it is OnDemand or other.

Reply:
OK, then let's strip the "service_checkin" prefix from the message.